File: packages/setpieces-py/tests/test_zones.py
import setpieces as sp
from setpieces.zones import classify_zone


def test_classify_zone_short():
    assert classify_zone(end_x=70.0, end_y=50.0, start_y=100.0) == "short"


def test_classify_zone_near_six():
    # start_y >= 50 mirrors end_y; end_y=30 stays 30 when start_y < 50
    assert classify_zone(end_x=96.0, end_y=30.0, start_y=0.0) == ("near", "six")


def test_classify_zone_central_edge_mirrored():
    # start_y >= 50 mirrors end_y -> my = 100 - 45 = 55 (central)
    assert classify_zone(end_x=88.0, end_y=45.0, start_y=100.0) == ("central", "edge")


def test_zone_breakdown_and_percentages(match):
    corners = sp.extract_set_pieces(match, kinds=["corner"])
    team_a_corners = [c for c in corners if c.team_id == "t_A"]
    counts = sp.zone_breakdown(team_a_corners)
    assert counts == {("central", "edge"): 1, ("near", "six"): 1}

    pct = sp.zone_percentages(team_a_corners)
    assert pct[("central", "edge")] == 50.0
    assert pct[("near", "six")] == 50.0

    team_b_corners = [c for c in corners if c.team_id == "t_B"]
    assert sp.zone_breakdown(team_b_corners) == {"short": 1}

File: packages/setpieces-r/setpieces/tests/testthat/test-zones.R
test_that("classify_zone: short delivery", {
  expect_equal(classify_zone(end_x = 70.0, end_y = 50.0, start_y = 100.0), "short")
})

test_that("classify_zone: near/six, no mirroring needed", {
  expect_equal(classify_zone(end_x = 96.0, end_y = 30.0, start_y = 0.0), "near_six")
})

test_that("classify_zone: central/edge, mirrored", {
  # start_y >= 50 mirrors end_y -> my = 100 - 45 = 55 (central)
  expect_equal(classify_zone(end_x = 88.0, end_y = 45.0, start_y = 100.0), "central_edge")
})

test_that("zone_breakdown and zone_percentages", {
  m <- sample_match()
  corners <- extract_set_pieces(m, kinds = "corner")
  team_a <- Filter(function(e) e$team_id == "t_A", corners)

  counts <- zone_breakdown(team_a)
  expect_equal(as.integer(counts[["central_edge"]]), 1L)
  expect_equal(as.integer(counts[["near_six"]]), 1L)

  pct <- zone_percentages(team_a)
  expect_equal(as.numeric(pct[["central_edge"]]), 50.0)
  expect_equal(as.numeric(pct[["near_six"]]), 50.0)

  team_b <- Filter(function(e) e$team_id == "t_B", corners)
  counts_b <- zone_breakdown(team_b)
  expect_equal(as.integer(counts_b[["short"]]), 1L)
})
